cirq/protocols/unitary.py

```diff
@@ -109,13 +139,22 @@ def has_unitary(val: Any) -> bool:
 
     Returns:
         If `val` has a _has_unitary_ method and its result is not
-        NotImplemented, that result is returned. Otherwise, if the value
-        has a _unitary_ method return if that has a non-default value.
-        Returns False if neither function exists.
+        NotImplemented, that result is returned. Otherwise, if `val` is a
+        cirq.Operation, a decomposition will be attempted and the resulting
+        unitary will be returned if unitaries exist for all operations of the
+        decompostion. Otherwise, if the value has a _unitary_ method return if
+        that has a non-default value. Returns False if neither function exists.
     """
     getter = getattr(val, '_has_unitary_', None)
     result = NotImplemented if getter is None else getter()
 
+    # Fallback to decomposition for operations
+    if (result is NotImplemented or not result) and isinstance(val, Operation):
```

Comment:
Just to make sure, you don't want the `if result is False`?
This causes an issue in the following way if `val` is a `GateOperation`:

1. We first check if `val` has the `_has_unitary_` method, which all instances of `GateOperation` do.
2. `GateOperation._has_unitary_` returns `has_unitary(self._gate)`.
3. If `_gate._has_unitary_` is not implemented, it will not try to decompose it since it is not a `GateOperation` anymore, and will therefore return `False`.
4. Therefore, we will not enter the decompose branch if we don't include  `if result is False`.

It might come down to this, does `has_unitary(val) == False` mean that a unitary doesn't exist, or that we just don't know what it is?

Reply:
Hmmm, I see. The fact that we're limiting ourselves to just operation for now is causing gate operation to not work, because it's implemented in terms of the gate.

I think the correct move here is to trust the False instead of trying to override it. It will start to work once we add support for gates. Maybe that indicates we should do it as part of this PR.